diff -ruN openssl-0.9.8h/build.bat openssl-0.9.8h-patched/build.bat
--- openssl-0.9.8h/build.bat	Thu Jan  1 02:00:00 1970
+++ openssl-0.9.8h-patched/build.bat	Thu Mar  5 14:24:48 2009
@@ -0,0 +1,23 @@
+call vcvarsall.bat x86
+perl Configure VC-WIN32
+call ms\do_ms debug
+nmake -f ms\nt.mak
+@if errorlevel 1 goto end
+call ms\do_masm
+nmake -f ms\nt.mak
+@if errorlevel 1 goto end
+call vcvarsall.bat x86_amd64
+perl Configure VC-WIN64A
+call ms\do_win64a debug
+nmake -f ms\nt.mak
+@if errorlevel 1 goto end
+call ms\do_win64a
+nmake -f ms\nt.mak
+@if errorlevel 1 goto end
+copy /y out32\*.lib ..\lib
+copy /y out32d\*.lib ..\lib
+copy /y out64\*.lib ..\lib
+copy /y out64d\*.lib ..\lib
+copy /y *.pdb ..\lib
+xcopy /e /i /y inc32\*.h ..\include
+:end
diff -ruN openssl-0.9.8h/crypto/perlasm/x86ms.pl openssl-0.9.8h-patched/crypto/perlasm/x86ms.pl
--- openssl-0.9.8h/crypto/perlasm/x86ms.pl	Fri May  2 02:11:32 2008
+++ openssl-0.9.8h-patched/crypto/perlasm/x86ms.pl	Thu Mar  5 14:47:52 2009
@@ -270,7 +270,6 @@
 
 	local($tmp)=<<"EOF";
 PUBLIC	_$func
-$extra
 _$func PROC NEAR
 	push	ebp
 	push	ebx
diff -ruN openssl-0.9.8h/ms/do_masm.bat openssl-0.9.8h-patched/ms/do_masm.bat
--- openssl-0.9.8h/ms/do_masm.bat	Thu Jul 19 20:39:06 2007
+++ openssl-0.9.8h-patched/ms/do_masm.bat	Thu Mar  5 14:51:24 2009
@@ -70,8 +70,8 @@
 echo on
 
 perl util\mkfiles.pl >MINFO
-perl util\mk1mf.pl VC-WIN32 >ms\nt.mak
-perl util\mk1mf.pl dll VC-WIN32 >ms\ntdll.mak
+perl util\mk1mf.pl %1 %2 %3 VC-WIN32 >ms\nt.mak
+perl util\mk1mf.pl dll %1 %2 %3 VC-WIN32 >ms\ntdll.mak
 
 perl util\mkdef.pl 32 libeay > ms\libeay32.def
 perl util\mkdef.pl 32 ssleay > ms\ssleay32.def
diff -ruN openssl-0.9.8h/ms/do_ms.bat openssl-0.9.8h-patched/ms/do_ms.bat
--- openssl-0.9.8h/ms/do_ms.bat	Tue May 17 03:07:14 2005
+++ openssl-0.9.8h-patched/ms/do_ms.bat	Thu Mar  5 01:59:05 2009
@@ -1,9 +1,9 @@
 
 perl util\mkfiles.pl >MINFO
-perl util\mk1mf.pl no-asm VC-WIN32 >ms\nt.mak
-perl util\mk1mf.pl dll no-asm VC-WIN32 >ms\ntdll.mak
-perl util\mk1mf.pl no-asm VC-CE >ms\ce.mak
-perl util\mk1mf.pl dll no-asm VC-CE >ms\cedll.mak
+perl util\mk1mf.pl no-asm %1 %2 %3 VC-WIN32 >ms\nt.mak
+perl util\mk1mf.pl dll no-asm %1 %2 %3 VC-WIN32 >ms\ntdll.mak
+perl util\mk1mf.pl no-asm %1 %2 %3 VC-CE >ms\ce.mak
+perl util\mk1mf.pl dll no-asm %1 %2 %3 VC-CE >ms\cedll.mak
 
 perl util\mkdef.pl 32 libeay > ms\libeay32.def
 perl util\mkdef.pl 32 ssleay > ms\ssleay32.def
diff -ruN openssl-0.9.8h/ms/do_nasm.bat openssl-0.9.8h-patched/ms/do_nasm.bat
--- openssl-0.9.8h/ms/do_nasm.bat	Thu Jul 19 20:39:06 2007
+++ openssl-0.9.8h-patched/ms/do_nasm.bat	Thu Mar  5 14:51:24 2009
@@ -71,9 +71,9 @@
 echo on
 
 perl util\mkfiles.pl >MINFO
-perl util\mk1mf.pl nasm VC-WIN32 >ms\nt.mak
-perl util\mk1mf.pl dll nasm VC-WIN32 >ms\ntdll.mak
-perl util\mk1mf.pl nasm BC-NT >ms\bcb.mak
+perl util\mk1mf.pl nasm %1 %2 %3 VC-WIN32 >ms\nt.mak
+perl util\mk1mf.pl dll nasm %1 %2 %3 VC-WIN32 >ms\ntdll.mak
+perl util\mk1mf.pl nasm %1 %2 %3 BC-NT >ms\bcb.mak
 
 perl util\mkdef.pl 32 libeay > ms\libeay32.def
 perl util\mkdef.pl 32 ssleay > ms\ssleay32.def
diff -ruN openssl-0.9.8h/ms/do_nt.bat openssl-0.9.8h-patched/ms/do_nt.bat
--- openssl-0.9.8h/ms/do_nt.bat	Sun May 23 02:24:38 1999
+++ openssl-0.9.8h-patched/ms/do_nt.bat	Thu Mar  5 01:59:05 2009
@@ -1,7 +1,7 @@
 
 perl util\mkfiles.pl >MINFO
-perl util\mk1mf.pl no-asm VC-NT >ms\nt.mak
-perl util\mk1mf.pl dll no-asm VC-NT >ms\ntdll.mak
+perl util\mk1mf.pl no-asm %1 %2 %3 VC-NT >ms\nt.mak
+perl util\mk1mf.pl dll no-asm %1 %2 %3 VC-NT >ms\ntdll.mak
 
 perl util\mkdef.pl libeay NT > ms\libeay32.def
 perl util\mkdef.pl ssleay NT > ms\ssleay32.def
diff -ruN openssl-0.9.8h/ms/do_win64a.bat openssl-0.9.8h-patched/ms/do_win64a.bat
--- openssl-0.9.8h/ms/do_win64a.bat	Tue Jul  5 02:24:12 2005
+++ openssl-0.9.8h-patched/ms/do_win64a.bat	Thu Mar  5 01:59:05 2009
@@ -2,8 +2,8 @@
 perl util\mkfiles.pl >MINFO
 perl ms\uplink.pl win64a > ms\uptable.asm
 ml64 -c -Foms\uptable.obj ms\uptable.asm
-perl util\mk1mf.pl no-asm VC-WIN64A >ms\nt.mak
-perl util\mk1mf.pl dll no-asm VC-WIN64A >ms\ntdll.mak
+perl util\mk1mf.pl no-asm %1 %2 %3 VC-WIN64A >ms\nt.mak
+perl util\mk1mf.pl dll no-asm %1 %2 %3 VC-WIN64A >ms\ntdll.mak
 
 perl util\mkdef.pl 32 libeay > ms\libeay32.def
 perl util\mkdef.pl 32 ssleay > ms\ssleay32.def
diff -ruN openssl-0.9.8h/ms/do_win64i.bat openssl-0.9.8h-patched/ms/do_win64i.bat
--- openssl-0.9.8h/ms/do_win64i.bat	Tue Jul  5 02:24:12 2005
+++ openssl-0.9.8h-patched/ms/do_win64i.bat	Thu Mar  5 01:59:05 2009
@@ -2,8 +2,8 @@
 perl util\mkfiles.pl >MINFO
 perl ms\uplink.pl win64i > ms\uptable.asm
 ias -o ms\uptable.obj ms\uptable.asm
-perl util\mk1mf.pl no-asm VC-WIN64I >ms\nt.mak
-perl util\mk1mf.pl dll no-asm VC-WIN64I >ms\ntdll.mak
+perl util\mk1mf.pl no-asm %1 %2 %3 VC-WIN64I >ms\nt.mak
+perl util\mk1mf.pl dll no-asm %1 %2 %3 VC-WIN64I >ms\ntdll.mak
 
 perl util\mkdef.pl 32 libeay > ms\libeay32.def
 perl util\mkdef.pl 32 ssleay > ms\ssleay32.def
diff -ruN openssl-0.9.8h/util/mk1mf.pl openssl-0.9.8h-patched/util/mk1mf.pl
--- openssl-0.9.8h/util/mk1mf.pl	Tue May 20 14:50:14 2008
+++ openssl-0.9.8h-patched/util/mk1mf.pl	Thu Mar  5 14:54:21 2009
@@ -188,8 +188,8 @@
 	$cflags.=' -DTERMIO';
 	}
 
-$out_dir=(defined($VARS{'OUT'}))?$VARS{'OUT'}:$out_def.($debug?".dbg":"");
-$tmp_dir=(defined($VARS{'TMP'}))?$VARS{'TMP'}:$tmp_def.($debug?".dbg":"");
+$out_dir=(defined($VARS{'OUT'}))?$VARS{'OUT'}:$out_def;
+$tmp_dir=(defined($VARS{'TMP'}))?$VARS{'TMP'}:$tmp_def;
 $inc_dir=(defined($VARS{'INC'}))?$VARS{'INC'}:$inc_def;
 
 $bin_dir=$bin_dir.$o unless ((substr($bin_dir,-1,1) eq $o) || ($bin_dir eq ''));
diff -ruN openssl-0.9.8h/util/pl/VC-32.pl openssl-0.9.8h-patched/util/pl/VC-32.pl
--- openssl-0.9.8h/util/pl/VC-32.pl	Fri Jan  4 02:40:00 2008
+++ openssl-0.9.8h-patched/util/pl/VC-32.pl	Thu Mar  5 14:51:54 2009
@@ -3,9 +3,6 @@
 # Win64 and WinCE [follow $FLAVOR variable to trace the differences].
 #
 
-$ssl=	"ssleay32";
-$crypto="libeay32";
-
 $o='\\';
 $cp='$(PERL) util/copy.pl';
 $mkdir='$(PERL) util/mkdir-p.pl';
@@ -28,14 +25,14 @@
     # per 0.9.8 release remaining warnings were explicitly examined and
     # considered safe to ignore.
     # 
-    $base_cflags=' /W3 /Gs0 /GF /Gy /nologo -DWIN32_LEAN_AND_MEAN -DL_ENDIAN -DDSO_WIN32 -DOPENSSL_SYSNAME_WIN32 -DOPENSSL_SYSNAME_WINNT -DUNICODE -D_UNICODE';
+    $base_cflags=' /Zi /W3 /Gs0 /GF /Gy /nologo -DWIN32_LEAN_AND_MEAN -DL_ENDIAN -DDSO_WIN32 -DOPENSSL_SYSNAME_WIN32 -DOPENSSL_SYSNAME_WINNT -DUNICODE -D_UNICODE';
     $base_cflags.=' -D_CRT_SECURE_NO_DEPRECATE';	# shut up VC8
     $base_cflags.=' -D_CRT_NONSTDC_NO_DEPRECATE';	# shut up VC8
-    my $f = $shlib?' /MD':' /MT';
     $lib_cflag='/Zl' if (!$shlib);	# remove /DEFAULTLIBs from static lib
-    $opt_cflags=$f.' /Ox';
-    $dbg_cflags=$f.'d /Od -DDEBUG -D_DEBUG';
-    $lflags="/nologo /subsystem:console /opt:ref";
+    $opt_cflags=' /MT /O2 /GL';
+    $dbg_cflags=' /MTd /Od -DDEBUG -D_DEBUG /RTC1 /GS';
+    $lflags="/nologo /subsystem:console /incremental:no";
+    $suffix='64';
     }
 elsif ($FLAVOR =~ /CE/)
     {
@@ -93,32 +90,37 @@
     }
 else	# Win32
     {
-    $base_cflags=' /W3 /WX /Gs0 /GF /Gy /nologo -DOPENSSL_SYSNAME_WIN32 -DWIN32_LEAN_AND_MEAN -DL_ENDIAN -DDSO_WIN32';
+    $base_cflags=' /Zi /W3 /WX /Gs0 /GF /Gy /nologo -DOPENSSL_SYSNAME_WIN32 -DWIN32_LEAN_AND_MEAN -DL_ENDIAN -DDSO_WIN32';
     $base_cflags.=' -D_CRT_SECURE_NO_DEPRECATE';	# shut up VC8
     $base_cflags.=' -D_CRT_NONSTDC_NO_DEPRECATE';	# shut up VC8
-    my $f = $shlib?' /MD':' /MT';
     $lib_cflag='/Zl' if (!$shlib);	# remove /DEFAULTLIBs from static lib
-    $opt_cflags=$f.' /Ox /O2 /Ob2';
-    $dbg_cflags=$f.'d /Od -DDEBUG -D_DEBUG';
-    $lflags="/nologo /subsystem:console /opt:ref";
+    $opt_cflags=' /MT /O2 /GL';
+    $dbg_cflags=' /MTd /Od -DDEBUG -D_DEBUG /RTC1 /GS';
+    $lflags="/nologo /subsystem:console /incremental:no";
+    $suffix='32';
     }
 $mlflags='';
 
-$out_def="out32"; $out_def.='_$(TARGETCPU)' if ($FLAVOR =~ /CE/);
-$tmp_def="tmp32"; $tmp_def.='_$(TARGETCPU)' if ($FLAVOR =~ /CE/);
-$inc_def="inc32";
-
 if ($debug)
 	{
 	$cflags=$dbg_cflags.$base_cflags;
-	$lflags.=" /debug";
+	$lflags.=" /debug /fixed:no";
 	$mlflags.=' /debug';
+	$suffix=$suffix.'d';
 	}
 else
 	{
 	$cflags=$opt_cflags.$base_cflags;
+	$lflags.=" /opt:ref /opt:icf /LTCG";
 	}
 
+$ssl=	"ssleay$suffix";
+$crypto="libeay$suffix";
+ 
+$out_def="out$suffix"; $out_def.='_$(TARGETCPU)' if ($FLAVOR =~ /CE/);
+$tmp_def="tmp$suffix"; $tmp_def.='_$(TARGETCPU)' if ($FLAVOR =~ /CE/);
+$inc_def="inc32";
+
 $obj='.obj';
 $ofile="/Fo";
 
@@ -139,7 +141,6 @@
 else
 	{
 	$ex_libs.=' gdi32.lib advapi32.lib user32.lib';
-	$ex_libs.=' bufferoverflowu.lib' if ($FLAVOR =~ /WIN64/);
 	}
 
 # As native NT API is pure UNICODE, our WIN-NT build defaults to UNICODE,
@@ -166,7 +167,7 @@
 	$asm=($ver gt $vew?"nasm":"nasmw")." -f win32";
 	$afile='-o ';
 } else {
-	$asm='ml /Cp /coff /c /Cx';
+	$asm='ml /Cp /coff /c /Cx /safeseh';
 	$asm.=" /Zi" if $debug;
 	$afile='/Fo';
 }
@@ -273,7 +274,6 @@
 		local($ex)=($target =~ /O_CRYPTO/)?'':' $(L_CRYPTO)';
 		if ($name eq "")
 			{
-			$ex.=' bufferoverflowu.lib' if ($FLAVOR =~ /WIN64/);
 			}
 		elsif ($FLAVOR =~ /CE/)
 			{
@@ -283,7 +283,6 @@
 			{
 			$ex.=' unicows.lib' if ($FLAVOR =~ /NT/);
 			$ex.=' wsock32.lib gdi32.lib advapi32.lib user32.lib';
-			$ex.=' bufferoverflowu.lib' if ($FLAVOR =~ /WIN64/);
 			}
 		$ex.=" $zlib_lib" if $zlib_opt == 1 && $target =~ /O_CRYPTO/;
 		$ret.="\t\$(LINK) \$(MLFLAGS) $efile$target $name @<<\n  \$(SHLIB_EX_OBJ) $objs $ex\n<<\n";
