diff -ruN fardev/docs/ClearPluginsCache.bat fardev-patched/docs/ClearPluginsCache.bat
--- fardev/docs/ClearPluginsCache.bat	Sat Oct 27 17:22:24 2007
+++ fardev-patched/docs/ClearPluginsCache.bat	Tue Mar 17 15:23:00 2009
@@ -1,7 +1,4 @@
 @echo off
 rem This batch file cleared the Plugin Cache
 
-echo REGEDIT4 > "%TEMP%\$DelOld$.reg"
-echo [-HKEY_CURRENT_USER\Software\Far\PluginsCache] >> "%TEMP%\$DelOld$.reg"
-start/wait regedit -s "%TEMP%\$DelOld$.reg"
-del "%TEMP%\$DelOld$.reg" > nul
+reg delete HKCU\Software\Far2\PluginsCache /f
diff -ruN fardev/docs/RestoreSettings.bat fardev-patched/docs/RestoreSettings.bat
--- fardev/docs/RestoreSettings.bat	Sat Oct 27 17:22:24 2007
+++ fardev-patched/docs/RestoreSettings.bat	Tue Mar 17 15:23:00 2009
@@ -2,17 +2,13 @@
 rem    This batch file restores Far settings from previously saved
 rem    files FarSave1.reg and FarSave2.reg to the registry
 
-if not exist FarSave?.reg goto import
+if not exist FarSave?.reg goto end
 
-echo REGEDIT4 > "%TEMP%\$DelOld$.reg"
-echo [-HKEY_CURRENT_USER\Software\Far] >> "%TEMP%\$DelOld$.reg"
-echo [-HKEY_LOCAL_MACHINE\Software\Far] >> "%TEMP%\$DelOld$.reg"
+reg delete HKCU\Software\Far2 /f
+reg delete HKLM\Software\Far2 /f
 
-start/wait regedit -s "%TEMP%\$DelOld$.reg"
-del "%TEMP%\$DelOld$.reg" > nul
+reg import FarSave1.reg
+reg import FarSave2.reg
+reg delete HKCU\Software\Far2\PluginsCache /f
 
-:import
-echo REGEDIT4 > "%TEMP%\$DelCache$.reg"
-echo [-HKEY_CURRENT_USER\Software\Far\PluginsCache] >> "%TEMP%\$DelCache$.reg"
-start/wait regedit -s FarSave1.reg FarSave2.reg "%TEMP%\$DelCache$.reg"
-del "%TEMP%\$DelCache$.reg" > nul
+:end
diff -ruN fardev/docs/SaveSettings.bat fardev-patched/docs/SaveSettings.bat
--- fardev/docs/SaveSettings.bat	Sat Oct 27 17:22:24 2007
+++ fardev-patched/docs/SaveSettings.bat	Tue Mar 17 15:23:00 2009
@@ -2,5 +2,5 @@
 rem    This batch file saves Far settings from the registry
 rem    to files FarSave1.reg and FarSave2.reg
 
-regedit /ea FarSave1.reg HKEY_CURRENT_USER\Software\Far
-regedit /ea FarSave2.reg HKEY_LOCAL_MACHINE\Software\Far
+reg export HKCU\Software\Far2 FarSave1.reg
+reg export HKLM\Software\Far2 FarSave2.reg
diff -ruN fardev/plugins/makefile_vc_def_inc fardev-patched/plugins/makefile_vc_def_inc
--- fardev/plugins/makefile_vc_def_inc	Mon Jan 12 16:40:23 2009
+++ fardev-patched/plugins/makefile_vc_def_inc	Tue Mar 17 15:23:00 2009
@@ -127,27 +127,27 @@
 NOWIN98=/OPT:NOWIN98
 !endif
 
+PDBNAME="$(OUTDIR)$(ADDOUTDIR)\$(NAME).pdb"
 !ifndef DEBUG
 CPP_OPT=/DNDEBUG /O1i
 !else
-PDBNAME="$(OUTDIR)$(ADDOUTDIR)\$(NAME).pdb"
-CPP_OPT=/DDEBUG /Od /Fd$(PDBNAME)
+CPP_OPT=/DDEBUG /Od
 !endif
 
 !ifdef AMD64
 COMMONLIB = $(COMMON)\libCRT64.lib
-CPP_PROJ_NO=/nologo /c /W3 /Gy /GF /Zp8 /J $(COMPAT64) /GS- /Gr /GR- /EHs-c- /LD /I"$(COMMON)" /I"$(COMINC)" $(ENV_INC_OPT) $(CPP_WIDE) /D_CRT_SECURE_NO_WARNINGS /D_CRT_NONSTDC_NO_WARNINGS /D_CRT_NON_CONFORMING_SWPRINTFS /D_WIN32_WINNT=0x0502 $(USERCPP)
+CPP_PROJ_NO=/nologo /c /W3 /Gy /GF /Zp8 /J $(COMPAT64) /GS- /Gr /GR- /EHs-c- /LD /I"$(COMMON)" /I"$(COMINC)" $(ENV_INC_OPT) $(CPP_WIDE) /D_CRT_SECURE_NO_WARNINGS /D_CRT_NONSTDC_NO_WARNINGS /D_CRT_NON_CONFORMING_SWPRINTFS /D_WIN32_WINNT=0x0502 $(USERCPP) /Zi /Fd$(PDBNAME)
 ULOUT=-Tpd+
 !elseif defined(IA64)
 COMMONLIB = $(COMMON)\libCRTIA64.lib
-CPP_PROJ_NO=/nologo /c /W3 /Gy /GF /Zp8 /J $(COMPAT64) /GS- /Gr /GR- /EHs-c- /LD /I"$(COMMON)" /I"$(COMINC)" $(ENV_INC_OPT) $(CPP_WIDE) /D_CRT_SECURE_NO_WARNINGS /D_CRT_NONSTDC_NO_WARNINGS /D_CRT_NON_CONFORMING_SWPRINTFS /D_WIN32_WINNT=0x0502 $(USERCPP)
+CPP_PROJ_NO=/nologo /c /W3 /Gy /GF /Zp8 /J $(COMPAT64) /GS- /Gr /GR- /EHs-c- /LD /I"$(COMMON)" /I"$(COMINC)" $(ENV_INC_OPT) $(CPP_WIDE) /D_CRT_SECURE_NO_WARNINGS /D_CRT_NONSTDC_NO_WARNINGS /D_CRT_NON_CONFORMING_SWPRINTFS /D_WIN32_WINNT=0x0502 $(USERCPP) /Zi /Fd$(PDBNAME)
 ULOUT=-Tpd+
 !else
 COMMONLIB = $(COMMON)\libCRT.lib
 !ifdef CPP_UNALIGN
 CPP_ALIGN=/Zp1
 !endif
-CPP_PROJ_NO=/nologo /c /W3 /Gy /GF $(CPP_ALIGN) /J /Gr /GS- /GR- /EHs-c- /LD /I"$(COMMON)" /I"$(COMINC)" $(ENV_INC_OPT) $(CPP_WIDE) /D_CRT_SECURE_NO_WARNINGS /D_CRT_NONSTDC_NO_WARNINGS /D_CRT_NON_CONFORMING_SWPRINTFS /D_WIN32_WINNT=0x0502 $(USERCPP)
+CPP_PROJ_NO=/nologo /c /W3 /Gy /GF $(CPP_ALIGN) /J /Gr /GS- /GR- /EHs-c- /LD /I"$(COMMON)" /I"$(COMINC)" $(ENV_INC_OPT) $(CPP_WIDE) /D_CRT_SECURE_NO_WARNINGS /D_CRT_NONSTDC_NO_WARNINGS /D_CRT_NON_CONFORMING_SWPRINTFS /D_WIN32_WINNT=0x0502 $(USERCPP) /Zi /Fd$(PDBNAME)
 ULOUT=-Tpd -Re
 !endif
 CPP_PROJ=$(CPP_PROJ_NO) /Fo"$(OBJDIR)\\"
@@ -172,15 +172,13 @@
 
 !ifndef ULINK
 LNK=$(_LINK_PATH_)link.exe
-LINK_DEBUG=/pdb:$(PDBNAME)
-CPP_DEBUG=/Zi
 !ifndef NEEDENTRY
 NOENTRY = /noentry
 !endif
 !ifndef NEEDDEFLIB
 NODEFLIB = /nodefaultlib
 !endif
-LINK_FLAGS=/nologo /dll /release /merge:.rdata=.text $(NOWIN98) $(NOENTRY) $(NODEFLIB) $(ENV_LIB_OPT) /def:"$(DEF)" /map:"$(MAP)" /out:"$(DLLFULLNAME)"
+LINK_FLAGS=/nologo /dll /release /merge:.rdata=.text $(NOWIN98) $(NOENTRY) $(NODEFLIB) $(ENV_LIB_OPT) /def:"$(DEF)" /map:"$(MAP)" /out:"$(DLLFULLNAME)" /debug /pdb:$(PDBNAME)
 !else
 LINK_DEBUG=-v
 CPP_DEBUG=/Z7
diff -ruN fardev/plugins/tmppanel/TmpClass.cpp fardev-patched/plugins/tmppanel/TmpClass.cpp
--- fardev/plugins/tmppanel/TmpClass.cpp	Tue Feb 24 01:24:39 2009
+++ fardev-patched/plugins/tmppanel/TmpClass.cpp	Tue Mar 17 14:58:13 2009
@@ -153,6 +153,17 @@
 #undef _NAME
 }
 
+bool IsSlash(TCHAR ch) {
+  return (ch == _T('\\')) || (ch == _T('/'));
+}
+
+bool IsAbsPath(const TCHAR* path) {
+  unsigned len = lstrlen(path);
+  if (len < 2) return false;
+  if ((path[1] == _T(':')) && ((len == 2) || IsSlash(path[2]))) return true;
+  else if (IsSlash(path[0]) && IsSlash(path[1])) return true;
+  else return false;
+}
 
 int TmpPanel::PutOneFile(PluginPanelItem &PanelItem)
 {
@@ -175,16 +186,17 @@
   if(CheckForCorrect(CurName,&CurPanelItem->FindData,Opt.AnyInPanel))
   {
     int NameOnly=(FSF.PointToName(CurName)==CurName);
+    bool IsRelPath=!IsAbsPath(CurName);
 #undef CurName
 
 #ifndef UNICODE
-    lstrcpy(CurPanelItem->FindData.cFileName, NameOnly ? CurDir : "");
+    lstrcpy(CurPanelItem->FindData.cFileName, IsRelPath ? CurDir : "");
     lstrcat(CurPanelItem->FindData.cFileName,PanelItem.FindData.cFileName);
 #else
     size_t wlen = lstrlen(PanelItem.FindData.lpwszFileName);
-    if (NameOnly) wlen += lstrlen(CurDir);
+    if (IsRelPath) wlen += lstrlen(CurDir);
     register wchar_t *wp = (wchar_t*)malloc((wlen+1)*sizeof(wchar_t));
-    lstrcpy(wp, NameOnly ? CurDir : L"");
+    lstrcpy(wp, IsRelPath ? CurDir : L"");
     lstrcat(wp, PanelItem.FindData.lpwszFileName);
     CurPanelItem->FindData.lpwszFileName = wp;
 #endif
@@ -275,6 +287,8 @@
 #ifdef UNICODE
       if(TmpPanelItem[i].FindData.lpwszFileName)
         TmpPanelItem[i].FindData.lpwszFileName = wcsdup(TmpPanelItem[i].FindData.lpwszFileName);
+      if(TmpPanelItem[i].FindData.lpwszAlternateFileName)
+        TmpPanelItem[i].FindData.lpwszAlternateFileName = wcsdup(TmpPanelItem[i].FindData.lpwszAlternateFileName);
 #endif
     }
   }
diff -ruN fardev/plugins/tmppanel/TmpMix.cpp fardev-patched/plugins/tmppanel/TmpMix.cpp
--- fardev/plugins/tmppanel/TmpMix.cpp	Tue Feb 24 01:24:39 2009
+++ fardev-patched/plugins/tmppanel/TmpMix.cpp	Tue Mar 17 14:55:25 2009
@@ -49,10 +49,10 @@
 #ifdef UNICODE
       if (Items[I].FindData.lpwszFileName)
         free ((wchar_t*)Items[I].FindData.lpwszFileName);
-/*
+
       if (Items[I].FindData.lpwszAlternateFileName)
         free ((wchar_t*)Items[I].FindData.lpwszAlternateFileName);
-*/
+
 #endif
     }
     free (Items);
diff -ruN fardev/unicode_far/findfile.cpp fardev-patched/unicode_far/findfile.cpp
--- fardev/unicode_far/findfile.cpp	Mon Mar 16 20:18:56 2009
+++ fardev-patched/unicode_far/findfile.cpp	Tue Mar 17 15:23:00 2009
@@ -909,8 +909,8 @@
   wchar_t *lpFileName = xf_wcsdup (PanelItem->FindData.lpwszFileName);
   wchar_t *lpFileNameShort = xf_wcsdup (PanelItem->FindData.lpwszAlternateFileName);
 
-  const wchar_t *lpFileNameToFind = PointToName(RemovePseudoBackSlash(lpFileName));
-  const wchar_t *lpFileNameToFindShort = PointToName(RemovePseudoBackSlash(lpFileNameShort));
+  const wchar_t *lpFileNameToFind = PointToName(lpFileName);
+  const wchar_t *lpFileNameToFindShort = PointToName(lpFileNameShort);
 
   if ( CtrlObject->Plugins.GetFindData (
       hPlugin,
@@ -925,11 +925,11 @@
       PluginPanelItem Item = *pItem;
 
       wchar_t *lpwszFileName = xf_wcsdup(NullToEmpty(pItem->FindData.lpwszFileName));
-      Item.FindData.lpwszFileName = xf_wcsdup (PointToName(RemovePseudoBackSlash(lpwszFileName)));
+      Item.FindData.lpwszFileName = xf_wcsdup (PointToName(lpwszFileName));
       xf_free (lpwszFileName);
 
       lpwszFileName = xf_wcsdup(NullToEmpty(pItem->FindData.lpwszAlternateFileName));
-      Item.FindData.lpwszAlternateFileName = xf_wcsdup (PointToName(RemovePseudoBackSlash(lpwszFileName)));
+      Item.FindData.lpwszAlternateFileName = xf_wcsdup (PointToName(lpwszFileName));
       xf_free (lpwszFileName);
 
       if ( !StrCmp (lpFileNameToFind, Item.FindData.lpwszFileName) &&
@@ -1700,8 +1700,7 @@
       string strFileName=FindList[FindExitIndex]->FindData.strFileName;
       Panel *FindPanel=CtrlObject->Cp()->ActivePanel;
 
-      if ((FindList[FindExitIndex]->ArcIndex != LIST_INDEX_NONE) &&
-          (!(ArcList[FindList[FindExitIndex]->ArcIndex]->Flags & OPIF_REALNAMES)))
+      if (FindList[FindExitIndex]->ArcIndex != LIST_INDEX_NONE)
       {
         HANDLE hPlugin = ArcList[FindList[FindExitIndex]->ArcIndex]->hPlugin;
         if (hPlugin == INVALID_HANDLE_VALUE)
@@ -1732,7 +1731,7 @@
               SearchMode==FFSEARCH_ALL ||
               SearchMode==FFSEARCH_ALL_BUTNETWORK ||
               SearchMode==FFSEARCH_INPATH)
-            CtrlObject->Plugins.SetDirectory(hPlugin,L"\\",OPM_FIND);
+            CtrlObject->Plugins.SetDirectory(hPlugin,L"\\",0);
 
           SetPluginDirectory(strFileName,hPlugin,TRUE);
         }
@@ -1805,32 +1804,19 @@
 {
   string strName;
   wchar_t *StartName,*EndName;
-  int IsPluginDir;
 
-  /* $ 19.01.2003 KM
-     Восстановлю поведение до 4 беты. Если в DirName есть
-     символ '\x1' значит это путь из плагина. Таким образом
-     легче определить плагиновые пути и, соответственно,
-     сделать правильный переход.
-  */
   strName = DirName;
-  if ( strName.GetLength() > 0 )
-    IsPluginDir=strName.Contains(L'\x1')?TRUE:FALSE;
-  else
-    IsPluginDir=FALSE;
 
   StartName = strName.GetBuffer();
-  while(IsPluginDir?(EndName=wcschr(StartName,L'\x1'))!=NULL:(EndName=wcschr(StartName,L'\\'))!=NULL)
+  if (*StartName==L'\\')
+  {
+    CtrlObject->Plugins.SetDirectory(hPlugin,L"\\",0);
+    StartName++;
+  }
+  while(EndName=wcschr(StartName,L'\\'))
   {
     *EndName=0;
-    // RereadPlugin
-    {
-      int FileCount=0;
-      PluginPanelItem *PanelData=NULL;
-      if (CtrlObject->Plugins.GetFindData(hPlugin,&PanelData,&FileCount,OPM_FIND))
-        CtrlObject->Plugins.FreeFindData(hPlugin,PanelData,FileCount);
-    }
-    CtrlObject->Plugins.SetDirectory(hPlugin,StartName,OPM_FIND);
+    CtrlObject->Plugins.SetDirectory(hPlugin,StartName,0);
     StartName=EndName+1;
   }
 
@@ -2294,8 +2280,6 @@
   string strPathName;
   strPathName = FullName;
 
-  RemovePseudoBackSlash(strPathName);
-
   CutToSlash(strPathName);
 
   if ( strPathName.IsEmpty() )
@@ -2702,7 +2686,8 @@
     struct OpenPluginInfo Info;
     CtrlObject->Plugins.GetOpenPluginInfo(hPlugin,&Info);
     strSaveDir = Info.CurDir;
-		strPluginSearchPath=strSaveDir+L"\x1";
+    strPluginSearchPath=Info.CurDir;
+    if (!strPluginSearchPath.IsEmpty()) AddEndSlash(strPluginSearchPath);
     WaitForSingleObject(hPluginMutex,INFINITE);
     if (SearchMode==FFSEARCH_ROOT ||
         SearchMode==FFSEARCH_ALL ||
@@ -2788,15 +2773,9 @@
       string strFullName;
       if (StrCmp(strCurName,L".")==0 || TestParentFolderName(strCurName))
         continue;
-      if (Flags & OPIF_REALNAMES)
-      {
-        strFullName = strCurName;
-      }
-      else
-      {
-        strFullName = strPluginSearchPath;
-        strFullName += strCurName;
-      }
+
+      strFullName = strPluginSearchPath;
+      strFullName += strCurName;
 
       /* $ 30.09.2003 KM
         Отфильтруем файлы не попадающие в действующий фильтр
@@ -2812,7 +2791,6 @@
           statusCS.Enter();
 
           strFindMessage = strFullName;
-          RemovePseudoBackSlash(strFindMessage);
           FindMessageReady=TRUE;
 
           statusCS.Leave();
@@ -2842,28 +2820,29 @@
         WaitForSingleObject(hPluginMutex,INFINITE);
 
 				size_t pos;
-        if (!strCurName.Contains(L'\x1') && CtrlObject->Plugins.SetDirectory(hPlugin,strCurName,OPM_FIND))
+        if (CtrlObject->Plugins.SetDirectory(hPlugin,strCurName,OPM_FIND))
         {
           ReleaseMutex(hPluginMutex);
 
           strPluginSearchPath += strCurName;
-          strPluginSearchPath += L"\x1";
+          strPluginSearchPath += L"\\";
           ScanPluginTree(hPlugin, Flags);
-          if (strPluginSearchPath.RPos(pos,L'\x1'))
+          if (strPluginSearchPath.RPos(pos,L'\\'))
             strPluginSearchPath.SetLength(pos);
+          if (strPluginSearchPath.RPos(pos,L'\\'))
+            strPluginSearchPath.SetLength(pos+1);
+          else
+            strPluginSearchPath.SetLength(0);
+
+          WaitForSingleObject(hPluginMutex,INFINITE);
+          if (!CtrlObject->Plugins.SetDirectory(hPlugin,L"..",OPM_FIND))
+            StopSearch=TRUE;
+          ReleaseMutex(hPluginMutex);
         }
-        if (strPluginSearchPath.RPos(pos,L'\x1'))
-          strPluginSearchPath.SetLength(pos+1);
         else
-          strPluginSearchPath.SetLength(0);
-        WaitForSingleObject(hPluginMutex,INFINITE);
-        if (!CtrlObject->Plugins.SetDirectory(hPlugin,L"..",OPM_FIND))
-          StopSearch=TRUE;
-        ReleaseMutex(hPluginMutex);
-        if (StopSearch) break;
+          ReleaseMutex(hPluginMutex);
       }
-      else
-        ReleaseMutex(hPluginMutex);
+      if (StopSearch) break;
     }
   }
   CtrlObject->Plugins.FreeFindData(hPlugin,PanelData,ItemCount);
@@ -3116,25 +3095,4 @@
   //SearchFromRoot[sz-1+MsgLenDiff]=0; //BUGBUG
 
   return strSearchFromRoot;
-}
-
-wchar_t *FindFiles::RemovePseudoBackSlash(wchar_t *FileName)
-{
-	for (int i=0;FileName[i]!=0;i++)
-	{
-		if (FileName[i]==L'\x1')
-			FileName[i]=L'\\';
-	}
-	return FileName;
-}
-
-string &FindFiles::RemovePseudoBackSlash(string &strFileName)
-{
-	wchar_t *lpwszFileName = strFileName.GetBuffer ();
-
-	RemovePseudoBackSlash (lpwszFileName);
-
-	strFileName.ReleaseBuffer ();
-
-	return strFileName;
 }
diff -ruN fardev/unicode_far/findfile.hpp fardev-patched/unicode_far/findfile.hpp
--- fardev/unicode_far/findfile.hpp	Sat Dec 27 17:00:42 2008
+++ fardev-patched/unicode_far/findfile.hpp	Tue Mar 17 15:23:00 2009
@@ -96,8 +96,6 @@
     static int GetPluginFile(DWORD ArcIndex, struct PluginPanelItem *PanelItem,
                              const wchar_t *DestPath, string &strResultName);
     static string &PrepareDriveNameStr(string &strSearchFromRoot,size_t sz);
-    static wchar_t *RemovePseudoBackSlash(wchar_t *FileName);
-    static string &RemovePseudoBackSlash(string &strFileName);
 
   public:
 
diff -ruN fardev/unicode_far/flupdate.cpp fardev-patched/unicode_far/flupdate.cpp
--- fardev/unicode_far/flupdate.cpp	Sun Feb 15 19:41:27 2009
+++ fardev-patched/unicode_far/flupdate.cpp	Tue Mar 17 15:23:00 2009
@@ -724,10 +724,6 @@
 
     //memset(CurListData,0,sizeof(*CurListData));
     PluginToFileListItem(CurPanelData,CurListData);
-    if(Info.Flags & OPIF_REALNAMES)
-    {
-        ConvertNameToShort (CurListData->strName, CurListData->strShortName);
-    }
     CurListData->Position=I;
 
     if ((Info.Flags & OPIF_USESORTGROUPS)/* && (CurListData->FileAttr & FILE_ATTRIBUTE_DIRECTORY)==0*/)
diff -ruN fardev/unicode_far/makefile_vc fardev-patched/unicode_far/makefile_vc
--- fardev/unicode_far/makefile_vc	Fri Mar 13 16:22:32 2009
+++ fardev-patched/unicode_far/makefile_vc	Tue Mar 17 15:23:00 2009
@@ -281,25 +281,25 @@
 	/D "_CRT_NONSTDC_NO_DEPRECATE"\
 	/D "_CRT_NON_CONFORMING_SWPRINTFS"
 
-CPP_PROJ_COMMON=$(MP) $(COMPAT64) /W3 /nologo $(FARSYSLOG) /EHs-c- $(DEFINES) /Gy /GF /Fo"$(INTDIR)\\" /Fd"$(INTDIR)\\" /J /c /FAcs /Fa"$(CODDIR)\\"
-CPP_PROJ_RELEASE=/MT /O2
-CPP_PROJ_DEBUG=/MTd /Od /Zi
+CPP_PROJ_COMMON=$(MP) $(COMPAT64) /W3 /nologo $(FARSYSLOG) /EHs-c- $(DEFINES) /Gy /GF /Fo"$(INTDIR)\\" /Fd"$(INTDIR)\\" /J /c /FAcs /Fa"$(CODDIR)\\" /Zi
+CPP_PROJ_RELEASE=/MT /O2 /GL
+CPP_PROJ_DEBUG=/MTd /Od
 
-LINK_COMMON=$(LINK32_LIBS) /OPT:REF /OPT:ICF $(NOWIN98) /nologo /subsystem:console /def:"$(DEF_FILE)" /out:"$(OUTDIR)\Far.exe" /map:"$(OUTDIR)\far.map" /pdb:"$(OUTDIR)\far.pdb" /release
+LINK_COMMON=$(LINK32_LIBS) $(NOWIN98) /nologo /subsystem:console /def:"$(DEF_FILE)" /out:"$(OUTDIR)\Far.exe" /map:"$(OUTDIR)\far.map" /pdb:"$(OUTDIR)\far.pdb" /release /debug /fixed:no
 
 !if "$(OUTDIR)" == "Release.32.vc"
 
 USEDEBUG=NDEBUG
 
 CPP_PROJ=$(CPP_PROJ_COMMON) $(CPP_PROJ_RELEASE) $(CPP_ADD_32) /Zp4
-LINK32_FLAGS=$(LINK_COMMON) /incremental:no /machine:i386
+LINK32_FLAGS=$(LINK_COMMON) /incremental:no /machine:i386 /OPT:REF /OPT:ICF /LTCG
 
 !elseif "$(OUTDIR)" == "Release.64.vc"
 
 USEDEBUG=NDEBUG
 
 CPP_PROJ=$(CPP_PROJ_COMMON) $(CPP_PROJ_RELEASE) /GS- /GR- /Zp8
-LINK32_FLAGS=$(LINK_COMMON) /incremental:no /machine:amd64
+LINK32_FLAGS=$(LINK_COMMON) /incremental:no /machine:amd64 /OPT:REF /OPT:ICF /LTCG
 ULINK_FLAGS=-Tpe+
 
 !elseif "$(OUTDIR)" == "Release.IA64.vc"
@@ -307,14 +307,14 @@
 USEDEBUG=NDEBUG
 
 CPP_PROJ=$(CPP_PROJ_COMMON) $(CPP_PROJ_RELEASE) /GS- /GR- /Zp8
-LINK32_FLAGS=$(LINK_COMMON) /incremental:no /machine:IA64
+LINK32_FLAGS=$(LINK_COMMON) /incremental:no /machine:IA64 /OPT:REF /OPT:ICF /LTCG
 
 !elseif "$(OUTDIR)" == "Debug.64.vc"
 
 USEDEBUG=_DEBUG
 
 CPP_PROJ=$(CPP_PROJ_COMMON) $(CPP_PROJ_DEBUG) /GS- /GR-
-LINK32_FLAGS=$(LINK_COMMON) /debug /machine:amd64
+LINK32_FLAGS=$(LINK_COMMON) /machine:amd64
 ULINK_FLAGS=-Tpe+ -v
 
 !elseif "$(OUTDIR)" == "Debug.IA64.vc"
@@ -322,14 +322,14 @@
 USEDEBUG=_DEBUG
 
 CPP_PROJ=$(CPP_PROJ_COMMON) $(CPP_PROJ_DEBUG) /GS- /GR-
-LINK32_FLAGS=$(LINK_COMMON) /debug /machine:IA64
+LINK32_FLAGS=$(LINK_COMMON) /machine:IA64
 
 !else
 
 USEDEBUG=_DEBUG
 
 CPP_PROJ=$(CPP_PROJ_COMMON) $(CPP_PROJ_DEBUG) $(CPP_ADD_32)
-LINK32_FLAGS=$(LINK_COMMON) /debug /machine:I386
+LINK32_FLAGS=$(LINK_COMMON) /machine:I386
 ULINK_FLAGS=-v
 
 !endif
